feat(curriculum): add merge sort project to Python curriculum (#52469)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-data-structures-by-building-the-merge-sort-algorithm/6569c2cbf6c993ea8cd85682.md

---
id: 6569c2cbf6c993ea8cd85682
title: Step 33
challengeType: 20
dashedName: step-33
---

# --description--

At this point, the `numbers` list has been sorted. Call the `print` function to print string `Sorted array: ` and the sorted list.

> To do that, concatenate `"Sorted array: "` and `str(numbers)` in the `print` function.

Reviewer: ```suggestion
To do that, concatenate `'Sorted array: '` and `str(numbers)` in the `print()` call.
```

With that, the merge sort algorithm is complete.

# --hints--

> You should have `print("sorted array: " + str(numbers))` in your code. Remember to add the space after the colon

Reviewer: ```suggestion
You should have `print('Sorted array: ' + str(numbers))` in your code. Remember to add the space after the colon
```

```js

> assert.match(code, /print\("sorted numbers: " \+ str\(\s*numbers\s*\)\)/)

Reviewer: ```suggestion
({ test: () => assert.match(e.code.original["main.py"], /^\s{4}print\s*\(\s*("|')Sorted array: \1\s*\+\s*str\s*\(\s*numbers\s*\)\s*\)/m) })
```

---
id: 6569c2cbf6c993ea8cd85682
title: Step 31
challengeType: 20
dashedName: step-31
---

# --description--

> At this point, the `numbers` list has been sorted. Let's print the sorted list.
> 
> Start by creating a `print()` function that prints string `Sorted array: ` and the the sorted list.

Reviewer: ```suggestion
At this point, the `numbers` list has been sorted. Call the `print` function to print string `Sorted array: ` and the sorted list.
```

> You can do this by concatenating `"Sorted array: "` and `str(numbers)` in the `print` function.
> 
> The `str` function converts the list to a string.

Reviewer: ```suggestion
To do that, concatenate `"Sorted array: "` and `str(numbers)` in the `print` function.

With that, the merge sort algorithm is complete.
```

# --hints--

> You should have `print("sorted array: " + str(numbers))`. Remember to add the space after the colon

Reviewer: ```suggestion
You should have `print("sorted array: " + str(numbers))` in your code. Remember to add the space after the colon.
```

```js
assert.match(code, /print\("sorted numbers: " \+ str\(\s*numbers\s*\)\)/)
```

# --seed--

## --seed-contents--

```py
def merge_sort(array):
    if len(array) <= 1:
        return
    
    middle_point = len(array) // 2
    left_part = array[:middle_point]
    right_part = array[middle_point:]

    merge_sort(left_part)
    merge_sort(right_part)

    left_array_index = 0
    right_array_index = 0
    sorted_index = 0

    while left_array_index < len(left_part) and right_array_index < len(right_part):
        if left_part[left_array_index] < right_part[right_array_index]:
            array[sorted_index] = left_part[left_array_index]
            left_array_index += 1
        else:
            array[sorted_index] = right_part[right_array_index]
            right_array_index += 1
        sorted_index += 1

    while left_array_index < len(left_part):
        array[sorted_index] = left_part[left_array_index]
        left_array_index += 1
        sorted_index += 1
    
    while right_array_index < len(right_part):
        array[sorted_index] = right_part[right_array_index]
        right_array_index += 1
        sorted_index += 1

> if __name__ == '__main__':
>     numbers = [4, 10, 6, 14, 2, 1, 8, 5]
> 
>     print("Unsorted array: ")
>     print(numbers)
> 
>     merge_sort(numbers)
> 
> --fcc-editable-region--
> 
> --fcc-editable-region--

Reviewer: ```suggestion
--fcc-editable-region--
if __name__ == '__main__':
    numbers = [4, 10, 6, 14, 2, 1, 8, 5]
    print("Unsorted array: ")
    print(numbers)    
    merge_sort(numbers)

--fcc-editable-region--
```

> ```

Reviewer: Please add the solution. See this as a reference: https://github.com/freeCodeCamp/freeCodeCamp/blob/4a365edfa6890afbdaacb2bc3b9ac38a0ed418c8/curriculum/challenges/english/20-upcoming-python/learn-recursion-by-solving-the-tower-of-hanoi-puzzle/657b6a4a49faec5c600287ba.md?plain=1#L58-L84
